wgpu-core/src/device/mod.rs

```diff
@@ -395,16 +401,20 @@ impl<B: GfxBackend> Device<B> {
         // TODO: 2D arrays, cubemap arrays
 
         let mut image = unsafe {
-            self.raw.create_image(
+            let mut image = self.raw.create_image(
                 kind,
                 desc.mip_level_count as hal::image::Level,
                 format,
                 hal::image::Tiling::Optimal,
                 usage,
                 view_capabilities,
-            )
-        }
-        .unwrap();
+            ).unwrap();
+            if !desc.label.is_null() {
+                let label = ffi::CStr::from_ptr(desc.label).to_string_lossy();
+                self.raw.set_image_name(&mut image, &label);
```

Comment:
hmm, that seems suboptimal. So we are getting `*const char`, then converting to `&str` to pass to gfx-rs, and then it probably converts it back to `*const char` at some point later on... Hmm.
At the same time, this is object creation, so it's not a big concern. Just that it would affect the API.
One way to address this (I'm not sure that we should, but still) is to have something like this:
```rust
enum Stringy<'a> {
  Rust(&'a str),
  C(&'a CStr),
}
```

Reply:
Yep - that's the issue I mentioned in the EDIT above.

Reply:
I think `gfx-hal` should just take ` CStr` and be done with it. It's purpose is to be low level.

Reply:
problem is that not all the backends agree on the strings they want. D3D, GL, and Vulkan do agree, but Metal wants NSString anyway, so it doesn't want CStr.

Reply:
You can convert a `CStr` to a `&str` by finding the length and creating the slice. You can't go in the other direction without allocating or having an intermediate buffer.

Reply:
but `str` is utf-8, and the byte stream of `CStr` may be wrong utf-8

Reply:
> problem is that not all the backends agree on the strings they want. D3D, GL, and Vulkan do agree, but Metal wants NSString anyway, so it doesn't want CStr.

Just convert the raw `*const c_char` to `NSString::stringWithUTF8String`, job done? :)

Reply:
It's a copy and convert step, just like one would do to convert `&str` to `NSString`, or `&str` to `CStr`. In that sense, `CStr` doesn't have an advantage.

Reply:
Ah, okay

Reply:
> It's a copy and convert step, just like one would do to convert &str to NSString, or &str to CStr. In that sense, CStr doesn't have an advantage.

Isn't there a copy with `&str` to `NSString` anyway? 

Wont `&str` -> `&CStr` -> `&str` -> `[&CStr or NSString]` always be more copies than  
`&str` -> `[&CStr or NSString]`.

Reply:
I don't think anybody proposed `&str -> &CStr -> &str -> [&CStr or NSString]`

Reply:
`&str` -> `&CStr` -> `&str` -> `[&CStr or NSString]`

This is what would currently happen with `wgpu-rs`. The label comes in as a `&str` from the rust descriptor. Then converted to a `*const c_char` to use with `wgpu-native` descriptor.  Then converted back to `&str` to pass to `gfx-hal`. Then gfx-hal converts it back to `CStr` to use with native API (or NSString in the mac case). 

Apologies, I'm using `CStr` and `*const c_char` interchangeably here to indicate a copy to null-terminated buffer.

Reply:
I see, yeah. This isn't intended. As outlined in https://github.com/gfx-rs/wgpu-rs/issues/156 , the main critical paths will be via wgpu-core (which works with `&str`) or websys, which needs its own strings anyway. Going via wgpu-native C API is kinda silly, only needed for compatibility with C apps and Dawn.